pyrasterframes/python/tests/PyRasterFramesTests.py

```diff
@@ -245,72 +333,189 @@ def test_exists_for_all(self):
+
+    def test_no_data_udf_handling(self):
+        t1 = Tile(np.array([[1, 2], [0, 4]]), CellType.uint8())
+        self.assertEqual(t1.cell_type.to_numpy_dtype(), np.dtype("uint8"))
+        e1 = Tile(np.array([[2, 3], [0, 5]]), CellType.uint8())
+        schema = StructType([StructField("tile", TileUDT(), False)])
+        df = self.spark.createDataFrame([{"tile": t1}], schema)
+
+        @udf(TileUDT())
+        def increment(t: Tile):
+            return t + 1
+
+        r1 = df.select(increment(df.tile).alias("inc")).first()["inc"]
+        self.assertEqual(r1, e1)
+
+
+class TileOps(TestEnvironment):
+
+    def test_addition(self):
+        t1 = Tile(np.array([[1, 2], [3, 4]]), CellType.int8().with_no_data_value(3))
+        e1 = np.ma.masked_equal(np.array([[5, 6], [7, 8]]), 7)
+        self.assertTrue(np.array_equal((t1 + 4).cells, e1))
+
+        t2 = Tile(np.array([[1, 2], [3, 4]]), CellType.int8().with_no_data_value(1))
+        e2 = np.ma.masked_equal(np.array([[3, 4], [3, 8]]), 3)
+        r2 = (t1 + t2).cells
+        self.assertTrue(np.ma.allequal(r2, e2))
+
+
+class PandasInterop(TestEnvironment):
+
+    def test_pandas_conversion(self):
+        import pandas as pd
+        #pd.options.display.max_colwidth = 256
+        cell_types = (ct for ct in rf_cell_types() if not (ct.is_raw() or ("bool" in ct.base_cell_type_name())))
+        tiles = [Tile(np.random.randn(5, 5) * 100, ct) for ct in cell_types]
+        in_pandas = pd.DataFrame({
+            'tile': tiles
         })
-        df = self.spark.createDataFrame(pandas_df)
-        df = df.withColumn("point_geom",
-                           st_point(df.x, df.y)
-                           )
-        df = df.withColumn("poly_geom", st_bufferPoint(df.point_geom, lit(1250.0)))
 
-        # Use python shapely UDT in a UDF
-        @udf("double")
-        def area_fn(g):
-            return g.area
+        in_spark = self.spark.createDataFrame(in_pandas)
+        out_pandas = in_spark.select(rf_identity('tile').alias('tile')).toPandas()
+        self.assertTrue(out_pandas.equals(in_pandas), str(in_pandas) + "\n\n" + str(out_pandas))
 
-        @udf("double")
-        def length_fn(g):
-            return g.length
+    def test_extended_pandas_ops(self):
+        import pandas as pd
 
-        df = df.withColumn("poly_area", area_fn(df.poly_geom))
-        df = df.withColumn("poly_len", length_fn(df.poly_geom))
+        self.assertIsInstance(self.rf.sql_ctx, SQLContext)
 
-        # Return UDT in a UDF!
-        def some_point(g):
-            return g.representative_point()
+        # Try to collect self.rf which is read from a geotiff
+        rf_collect = self.rf.take(2)
+        self.assertTrue(
+            all([isinstance(row.tile.cells, np.ndarray) for row in rf_collect]))
 
-        some_point_udf = udf(some_point, PointUDT())
+        # Try to create a tile from numpy.
+        self.assertEqual(Tile(np.random.randn(10, 10), CellType.int8()).dimensions(), [10, 10])
```

Comment:
it is not material to the test case  but in case someone copies code out of here 

```suggestion
        self.assertEqual(Tile(np.random.randint(0, 127, (10, 10), 'int8'), CellType.int8()).dimensions(), [10, 10])
```